### ISSUE TITLE
Variable substitution not working when there's whitespace between the braces and the variable

### ISSUE THREAD

**Issue by cgorman** [argoproj/argo-workflows#4484]
## Summary 

Per [the documentation](https://argoproj.github.io/argo/variables/), we should be able to write

```yaml
    # args: ["{{ inputs.parameters.message }}"]       <- good
      args: ["{{inputs.parameters.message}}"]         #  good
```
but including whitespace as the first example does passes the string literal rather than substituting the variable.

## Diagnostics

**What Kubernetes provider are you using?** 
Microk8s on WIndows

**What version of Argo Workflows are you running?** 
2.11.7

Note that this is the same as the global parameters test (because that's what I was having trouble with originally) except in the args for argosay I added spaces between the braces of the first three variables
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  labels:
    workflows.argoproj.io/completed: "true"
    workflows.argoproj.io/creator: system-serviceaccount-test-argo-server
    workflows.argoproj.io/phase: Succeeded
  name: global-parameters-complex-d9nj7
  namespace: test
  resourceVersion: "613667"
  selfLink: /apis/argoproj.io/v1alpha1/namespaces/test/workflows/global-parameters-complex-d9nj7
  uid: eac9cc89-b5c2-40e0-8ee6-8aa48dd4978e
spec:
  arguments:
    parameters:
    - name: message1
      value: hello world
    - name: message2
      value: foo bar
  entrypoint: test-workflow
  templates:
  - arguments: {}
    inputs:
      parameters:
      - name: message1
      - name: message-internal
        value: '{{workflow.parameters.message1}}'
    metadata: {}
    name: test-workflow
    outputs: {}
    steps:
    - - arguments:
          parameters:
          - name: message1
            value: world hello
          - name: message2
            value: '{{inputs.parameters.message1}}'
          - name: message3
            value: '{{workflow.parameters.message2}}'
          - name: message4
            value: '{{inputs.parameters.message-internal}}'
        name: step1
        template: whalesay2
  - arguments: {}
    container:
      args:
      - 'Global 1: {{ workflow.parameters.message1 }} Input 1: {{ inputs.parameters.message1
        }} Input 2/Steps Input 1/Global 1: {{ inputs.parameters.message2 }} Input
        3/Global 2: {{inputs.parameters.message3}} Input4/Steps Input 2 internal/Global
        1: {{inputs.parameters.message4}}'
      command:
      - cowsay
      image: argoproj/argosay:v1
      name: ""
      resources: {}
    inputs:
      parameters:
      - name: message1
      - name: message2
      - name: message3
      - name: message4
    metadata: {}
    name: whalesay2
    outputs: {}
```

```
time="2020-11-06T19:27:15Z" level=info msg="Processing workflow" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:15Z" level=info msg="Updated phase  -> Running" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:15Z" level=info msg="Steps node global-parameters-complex-d9nj7 initialized Running" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:15Z" level=info msg="StepGroup node global-parameters-complex-d9nj7-697151445 initialized Running" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:15Z" level=info msg="Pod node global-parameters-complex-d9nj7-1800787214 initialized Pending" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:15Z" level=info msg="Created pod: global-parameters-complex-d9nj7[0].step1 (global-parameters-complex-d9nj7-1800787214)" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:15Z" level=info msg="Workflow step group node global-parameters-complex-d9nj7-697151445 not yet completed" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:15Z" level=info msg="Workflow update successful" namespace=test phase=Running resourceVersion=613645 workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:16Z" level=info msg="Processing workflow" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:16Z" level=info msg="Updating node global-parameters-complex-d9nj7-1800787214 message: ContainerCreating"
time="2020-11-06T19:27:16Z" level=info msg="Skipped pod global-parameters-complex-d9nj7[0].step1 (global-parameters-complex-d9nj7-1800787214) creation: already exists" namespace=test podPhase=Pending workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:16Z" level=info msg="Workflow step group node global-parameters-complex-d9nj7-697151445 not yet completed" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:16Z" level=info msg="Workflow update successful" namespace=test phase=Running resourceVersion=613651 workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:17Z" level=info msg="Processing workflow" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:17Z" level=info msg="Setting node global-parameters-complex-d9nj7-1800787214 outputs"
time="2020-11-06T19:27:17Z" level=info msg="Updating node global-parameters-complex-d9nj7-1800787214 status Pending -> Running"
time="2020-11-06T19:27:17Z" level=info msg="Workflow step group node global-parameters-complex-d9nj7-697151445 not yet completed" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:17Z" level=info msg="Workflow update successful" namespace=test phase=Running resourceVersion=613664 workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="Processing workflow" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="Updating node global-parameters-complex-d9nj7-1800787214 status Running -> Succeeded"
time="2020-11-06T19:27:18Z" level=info msg="Step group node &NodeStatus{ID:global-parameters-complex-d9nj7-697151445,Name:global-parameters-complex-d9nj7[0],DisplayName:[0],Type:StepGroup,TemplateName:test-workflow,TemplateRef:nil,Phase:Running,BoundaryID:global-parameters-complex-d9nj7,Message:,StartedAt:2020-11-06 19:27:15 +0000 UTC,FinishedAt:0001-01-01 00:00:00 +0000 UTC,PodIP:,Daemoned:nil,Inputs:nil,Outputs:nil,Children:[global-parameters-complex-d9nj7-1800787214],OutboundNodes:[],StoredTemplateID:,WorkflowTemplateName:,TemplateScope:local/global-parameters-complex-d9nj7,ResourcesDuration:ResourcesDuration{},HostNodeName:,MemoizationStatus:nil,SynchronizationStatus:nil,} successful" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="node global-parameters-complex-d9nj7-697151445 phase Running -> Succeeded" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="node global-parameters-complex-d9nj7-697151445 finished: 2020-11-06 19:27:18.794235409 +0000 UTC" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="Outbound nodes of global-parameters-complex-d9nj7-1800787214 is [global-parameters-complex-d9nj7-1800787214]" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="Outbound nodes of global-parameters-complex-d9nj7 is [global-parameters-complex-d9nj7-1800787214]" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="node global-parameters-complex-d9nj7 phase Running -> Succeeded" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="node global-parameters-complex-d9nj7 finished: 2020-11-06 19:27:18.794414407 +0000 UTC" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="Checking daemoned children of global-parameters-complex-d9nj7" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="Updated phase Running -> Succeeded" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="Marking workflow completed" namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="Checking daemoned children of " namespace=test workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:18Z" level=info msg="Workflow update successful" namespace=test phase=Succeeded resourceVersion=613667 workflow=global-parameters-complex-d9nj7
time="2020-11-06T19:27:19Z" level=info msg="Labeled pod completed" namespace=test pod=global-parameters-complex-d9nj7-1800787214
```

Please either fix or update the documentation, I spent a week chasing a bug that didn't exist because of this.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by sarabala1979** [comment]
Would you like to fix the document and submit the PR?

**Comment by simster7** [comment]
I'll take a look

**Comment by simster7** [comment]
I am so far unable to reproduce this issue with the example from the variables doc or the workflow you provided:

```
$ cat ex.yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: hello-world-parameters-
spec:
  entrypoint: whalesay
  arguments:
    parameters:
    - name: message
      value: hello world
  templates:
  - name: whalesay
    inputs:
      parameters:
      - name: message
    container:
      image: docker/whalesay
      command: [cowsay]
      args: ["{{ inputs.parameters.message }}"]
      # args: ["{{inputs.parameters.message}}"]

$ argo submit --watch  ex.yaml

...

Name:                hello-world-parameters-ptg7p
Namespace:           argo
ServiceAccount:      default
Status:              Succeeded
Conditions:
 Completed           True
Created:             Mon Nov 09 10:49:18 -0600 (5 seconds ago)
Started:             Mon Nov 09 10:49:18 -0600 (5 seconds ago)
Finished:            Mon Nov 09 10:49:23 -0600 (now)
Duration:            5 seconds
Progress:            1/1
ResourcesDuration:   2s*(1 cpu),1s*(100Mi memory)
Parameters:
  message:           hello world

STEP                             TEMPLATE  PODNAME                       DURATION  MESSAGE
 ✔ hello-world-parameters-ptg7p  whalesay  hello-world-parameters-ptg7p  3s
$ kubectl logs hello-world-parameters-ptg7p -c main
 _____________
< hello world >
 -------------
    \
     \
      \
                    ##        .
              ## ## ##       ==
           ## ## ## ##      ===
       /""""""""""""""""___/ ===
  ~~~ {~~ ~~~~ ~~~ ~~~~ ~~ ~ /  ===- ~~~
       \______ o          __/
        \    \        __/
          \____\______/
```

**Comment by simster7** [comment]
Do you have any more info that could help me diagnose?

**Comment by cgorman** [comment]
Wow that's wild! I have no idea then, the only things I can think of are that I'm using a community helm chart for deployment and I'm running it on a windows multipass microk8s setup. Any of those jump out to you as the culprit?

**Comment by simster7** [comment]
We have had some issues with Argo variables and helm in the past. Take a look at this thread for more info: https://github.com/argoproj/argo/issues/2430

Let me know if you need more help

**Comment by simster7** [comment]
@cgorman Quick update:

I identified the PR responsible for the fix: #4310. It was merged while I was OOO, so I wasn't aware of it. I have backported it to 2.11 and it will be in the next release.

**Comment by cgorman** [comment]
Awesome, thanks!
